keps/sig-storage/20191012-storagepool-advplacement.md

```diff
@@ -0,0 +1,364 @@
+
+The placement choices in this use case cannot be expressed using existing Kubernetes mechanism.
+
+![Story2-LocalDrives](./kep-storagepool-story2.png)
+
+Beyond the existence of the Placement Choice / Storage Pool, Minio further uses metadata to know that the storage pool has enough capacity to accommodate the desired PVC. If for example physical disks have different size, then Minio may decide to create PVCs of different sizes to still work in this scenario.
+
+#### Story 3
+
+Let’s assume a sharded and replicated MongoDB instance. Let’s further assume that the Cloud Storage Service used for the PVCs is comprised of 3 distinct Storage-Pods (not to be confused with K8s Pods) in both of the two available AZs, such that the 3 Storage-Pods fail independently by having dedicated power, networking, racks, hosts and disks. The CSI driver of the Cloud Storage service announces that 3 Storage-Pods per AZ as placement choices (AZ1-SP1, AZ1-SP2, AZ1-SP3, AZ2-SP1, etc.) and also advertises that all nodes in AZ1 can access all Storage-Pods in AZ1 (as a simple example). Let’s further assume MongoDB is configured for 3-way replication per shard. MongoDB decides to create 3 Pods for a given shard:
+* Pod1 has an affinity to AZ1, and its PVCs have an affinity to AZ1-SP1.
+* Pod2 has an affinity to AZ1, and its PVCs have an affinity to AZ1-SP2.
+* Pod3 has an affinity to AZ2, and its PVCs have an affinity to AZ2-SP1.
+
+If we compare this to the case where we can’t express the PVC affinity to the Storage-Pod, Pod1 and Pod2 may land on the same Storage-Pod and so if that Storage-Pod fails, it would take down two Pods. In the scenario described here, the math will show a higher availability and durability.
+
+The placement choices in this use case cannot be expressed using existing Kubernetes mechanism.
+
+![Story3-MultiStoragePods](./kep-storagepool-story3.png)
+
+
+### Implementation Details/Notes/Constraints
+
+#### New CRD to surface Storage Pools
+```
+apiVersion: storagepool.storage.k8s.io/v1alpha1
+kind: StoragePool
+metadata:
+  name: storagePool1
+  labels:
+     labelKey: labelValue
+spec:
+  driver: myDriver
+  parameters:
+     csiBackendSpecificOpaqueKey: csiBackendSpecificOpaqueValue
```

Comment:
The section "Using StorageClass for placement" discusses that applications should copy these into the StorageClass, until a separate KEP automates placement.

Reply:
Sorry, somehow I missed that.

Which binding mode is that StorageClass meant to have? For the intended use case it'll probably be the normal immediate binding mode, right?

With delayed binding mode you would be back at the problem described in https://github.com/kubernetes/enhancements/pull/1353: a node might get chosen without Kubernetes knowing whether the node is suitable for the intended StoragePool.

Reply:
That is correct, let me say that more explicitly. I added "Most commonly this will lead to a "compute-follows-storage" design, i.e. it won't use `WaitForFirstConsumer` and instead have the PV be created first, and then have the Pod follow it based on which nodes the volume is accessible from."